internal/api/v1/application/stage.go

```diff
@@ -825,3 +852,139 @@ func mountRegistryCerts(app stageParam, volumes []corev1.Volume, volumeMounts []
 func appendEnvVar(envs []corev1.EnvVar, name, value string) []corev1.EnvVar {
 	return append(envs, corev1.EnvVar{Name: name, Value: value})
 }
+
+type StagingScriptConfig struct {
+	Name          string // config name. Needed to mount the resource in the pod
+	Builder       string // glob pattern for builders supported by this resource
+	UserID        int64  // user id to run the build phase with (`cnb` user)
+	GroupID       int64  // group id to run the build hase with
+	Base          string // optional, name of resource to pull the other parts from
+	DownloadImage string // image to run the download phase with
+	UnpackImage   string // image to run the unpack phase with
+}
+
+func DetermineStagingScripts(ctx context.Context,
+	logger logr.Logger,
+	cluster *kubernetes.Cluster,
+	namespace, builder string) (*StagingScriptConfig, error) {
+
+	logger.Info("locate staging scripts", "namespace", namespace, "builder", builder)
+
+	configmapSelector := labels.Set(map[string]string{
+		"app.kubernetes.io/component": "epinio-staging",
+	}).AsSelector().String()
+
+	configmapList, err := cluster.Kubectl.CoreV1().ConfigMaps(namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: configmapSelector,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("locate staging scripts", "possibles", len(configmapList.Items))
+
+	var candidates []*StagingScriptConfig
+	for _, configmap := range configmapList.Items {
+		config, err := NewStagingScriptConfig(configmap)
+		if err != nil {
+			return nil, err
+		}
+		candidates = append(candidates, config)
+	}
+
+	// Sort the candidates by name to have a deterministic search order
+	slices.SortFunc(candidates, SortByName)
+
+	var fallback *StagingScriptConfig
+
+	for _, config := range candidates {
+		pattern := config.Builder
+
+		logger.Info("locate staging scripts - checking", "name", config.Name, "for", pattern)
+		if pattern == "*" || pattern == "" {
```

Comment:
Now that the scripts are sorted I guess we can safely drop the `"*"` from the fallback. It will be the first to match, and so it will return immediately after this.

https://play.golang.com/p/Ts6mRidtGJG
```suggestion
		if pattern == "" {
```

Reply:
It will be nice to log also the ordered scripts, to debug what happened if a different one was picked up.

Reply:
I want `*` to match __last__. As a fallback having it preempt everything else is the wrong action.

Reply:
> It will be nice to log also the ordered scripts, to debug what happened if a different one was picked up.

Grep the server log for `staging` to see the match activity

Example:

```
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "namespace"="epinio" "builder"="paketobuildpacks/builder-jammy-full:0.3.290"
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "possibles"=3
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts - checking" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "name"="epinio-stage-scripts" "for"="*"
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts - saving fallback" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4"
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts - checking" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "name"="epinio-stage-scripts-bionic" "for"="paketobuildpacks/builder:*"
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts - checking" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "name"="epinio-stage-scripts-jammy" "for"="paketobuildpacks/builder-jammy-*:*"
2023/10/20 07:52:45 EpinioServer: "level"=1 "msg"="locate staging scripts - match" "requestId"="fb0b15fa-746d-4c0b-aa22-6592a432c7b4" "name"="epinio-stage-scripts-jammy" "for"="paketobuildpacks/builder-jammy-*:*" "builder"="paketobuildpacks/builder-jammy-full:0.3.290"
```